datafusion/src/sql/planner.rs

```diff
@@ -195,26 +195,42 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
             } => {
                 let left_plan = self.set_expr_to_plan(left.as_ref(), None, ctes)?;
                 let right_plan = self.set_expr_to_plan(right.as_ref(), None, ctes)?;
+                let join_keys = left_plan
```

Comment:
i am a bit concerned about the extra overhead here for the union branches. perhaps also relevant to https://github.com/apache/arrow-datafusion/pull/1258/files#r744178212

Reply:
I add an `if`, only `intersect` and `except` will get `join_keys`